File: src/main/java/application/gameplay/PlayDecision.java
package application.gameplay;

import model.basic.Tile;

public class PlayDecision {
    public enum Action {
        HU,
        CHOW,
        PUNG,
        KONG,
        SKIP,
        DISCARD,
        NONE
    }

    private final Action action;
    private final Tile tile;
    private final String reason;
    private final String source;

    public PlayDecision(Action action, Tile tile, String reason, String source) {
        this.action = action;
        this.tile = tile;
        this.reason = reason == null ? "" : reason;
        this.source = source == null ? "" : source;
    }

    public static PlayDecision of(Action action, String reason, String source) {
        return new PlayDecision(action, null, reason, source);
    }

    public static PlayDecision discard(Tile tile, String reason, String source) {
        return new PlayDecision(Action.DISCARD, tile, reason, source);
    }

    public static PlayDecision none(String reason, String source) {
        return new PlayDecision(Action.NONE, null, reason, source);
    }

    public Action getAction() {
        return action;
    }

    public Tile getTile() {
        return tile;
    }

    public String getReason() {
        return reason;
    }

    public String getSource() {
        return source;
    }
}

File: UnityVersion/Assets/Scripts/Core/Gameplay/PlayDecision.cs
using SichuanMahjong.Core.Model;

namespace SichuanMahjong.Core.Gameplay
{
    public class PlayDecision
    {
        public enum ActionType
        {
            HU,
            CHOW,
            PUNG,
            KONG,
            SKIP,
            DISCARD,
            NONE
        }

        private readonly ActionType action;
        private readonly Tile tile;
        private readonly string reason;
        private readonly string source;

        public PlayDecision(ActionType action, Tile tile, string reason, string source)
        {
            this.action = action;
            this.tile = tile;
            this.reason = reason ?? "";
            this.source = source ?? "";
        }

        public static PlayDecision Of(ActionType action, string reason, string source)
        {
            return new PlayDecision(action, null, reason, source);
        }

        public static PlayDecision Discard(Tile tile, string reason, string source)
        {
            return new PlayDecision(ActionType.DISCARD, tile, reason, source);
        }

        public static PlayDecision None(string reason, string source)
        {
            return new PlayDecision(ActionType.NONE, null, reason, source);
        }

        public ActionType GetAction()
        {
            return action;
        }

        public Tile GetTile()
        {
            return tile;
        }

        public string GetReason()
        {
            return reason;
        }

        public string GetSource()
        {
            return source;
        }
    }
}
